### ISSUE TITLE
M1 Mac appear not to run databuilder correctly

### ISSUE THREAD

**Issue by CarolineMorton** [opensafely-core/databuilder#538]
Just a note for the future that I cannot run tests on databuilder on a M1 mac. Unit tests work fine. The docker container for MSSQL starts and then immediately stops. I have not explored this in great detail as I have the option of switching to an intel. 

Next steps:
- Other mac m1 users try to run tests that require docker to see if problem persist
- If they do, change documentation to indicate this

**Comment by StevenMaude** [comment]
The MS SQL images are [for Linux](https://hub.docker.com/_/microsoft-mssql-server).

Does `export DOCKER_DEFAULT_PLATFORM=linux/amd64` or similar help here?

**Comment by evansd** [comment]
Yes, this is the problem. It would be worth trying to run this command and seeing what happens:
```
docker run --rm \
 --platform linux/amd64 \
 -e ACCEPT_EULA=y \
 mcr.microsoft.com/mssql/server:2017-CU25-ubuntu-16.04
```

All being well you'll see a load of log output from the database booting up, finishing with something like:
```
2022-06-23 14:54:18.88 spid5s      Recovery is complete. This is an informational message only. No user action is required.
2022-06-23 14:54:18.90 spid24s     The default language (LCID 0) has been set for engine and full-text services.
```

However the Intel emulation this uses is definitely [not perfect](https://docs.docker.com/desktop/mac/apple-silicon/#known-issues) and it may well fail, in which case there's not an awful lot we can do at this stage — beyond not running the MSSQL tests locally on M1s.

**Comment by remlapmot** [comment]
(I tried this because several of the Bristol team have M1 Macs) 

The example fails for me - but I found a solution by searching for the error message I got as summarised [here](https://github.com/microsoft/mssql-docker/issues/668#issuecomment-1149840353) which uses [Colima](https://github.com/abiosoft/colima#readme) as follows

```bash
brew install colima
colima start --arch amd
```

The `docker run ... ` example then worked as described.

(Once finished run) 
```bash
colima stop
```

Whether this would be enough to make these work within databuilder or whether some additional tweak would be needed, I don't know. I sense not running these tests on M1s is the easier fix.

**Comment by evansd** [comment]
Oh thanks Tom, that's very helpful. I'd not encountered [Colima](https://github.com/abiosoft/colima) before.

If the MSSQL container boots successfully that's _reasonable_ grounds for thinking it might all just work, which would be lovely.

And just to be clear (on what I'm pretty sure you've understood anyway) that this is only relevant for developers of databuilder who need to run tests against MSSQL. I don't anticipate (fingers crossed!) any issues for _users_ of databuilder on M1s.

**Comment by holly-cummins** [comment]
Even the Azure SQL Edge server seems to have regressed on M1. With `:latest`, I get the same segfault:

```
11:34:50.000 MSSQL:Azure SQL Edge will run as non-root by default.
11:34:50.000 MSSQL:This container is running as user mssql.
11:34:50.000 MSSQL:To learn more visit https://go.microsoft.com/fwlink/?linkid=2140520.
11:34:50.000 MSSQL:2022/07/01 10:34:50 [launchpadd] INFO: Extensibility Log Header: <timestamp> <process> <sandboxId> <sessionId> <message>
11:34:50.000 MSSQL:2022/07/01 10:34:50 [launchpadd] WARNING: Failed to load /var/opt/mssql/mssql.conf ini file with error open /var/opt/mssql/mssql.conf: no such file or directory
11:34:50.000 MSSQL:2022/07/01 10:34:50 [launchpadd] INFO: DataDirectories =  /bin:/etc:/lib:/lib32:/lib64:/sbin:/usr/bin:/usr/include:/usr/lib:/usr/lib32:/usr/lib64:/usr/libexec/gcc:/usr/sbin:/usr/share:/var/lib:/opt/microsoft:/opt/mssql-extensibility:/opt/mssql/mlservices:/opt/mssql/lib/zulu-jre-11:/opt/mssql-tools
11:34:50.000 MSSQL:2022/07/01 10:34:50 Drop permitted effective capabilities.
11:34:50.000 MSSQL:/opt/mssql/bin/sqlservr: Invalid mapping of address 0x40092f3000 in reserved address space below 0x400000000000. Possible causes:
11:34:50.000 MSSQL:1) the process (itself, or via a wrapper) starts-up its own running environment sets the stack size limit to unlimited via syscall setrlimit(2);
11:34:50.000 MSSQL:2) the process (itself, or via a wrapper) adjusts its own execution domain and flag the system its legacy personality via syscall personality(2);
11:34:50.000 MSSQL:3) sysadmin deliberately sets the system to run on legacy VA layout mode by adjusting a sysctl knob vm.legacy_va_layout.
11:34:50.000 MSSQL:
```

Using `mcr.microsoft.com/azure-sql-edge:1.0.6` seems to work, though.

**Comment by remlapmot** [comment]
Docker 4.16.0 was released yesterday. It has a new Beta feature for Apple silicon Macs - Rosetta for Linux, [release notes](https://docs.docker.com/desktop/release-notes/#4160). It's off by default - once enabled

<img width="1270" alt="screen-1" src="https://user-images.githubusercontent.com/3777473/212493110-cb6aba11-afb5-4a69-b15e-adef85ce4e6f.png">

this example code runs for me
```
docker run --rm \
 --platform linux/amd64 \
 -e ACCEPT_EULA=y \
 mcr.microsoft.com/mssql/server:2017-CU25-ubuntu-16.04
```

You can see another user reporting similarly, https://github.com/microsoft/mssql-docker/issues/734#issuecomment-1382358649

**Comment by StevenMaude** [comment]
@remlapmot: thanks for adding the note about Rosetta :+1: 

If someone can get the Data Builder test suite running with Rosetta, then I think it's sufficient to document this solution in `DEVELOPERS.md` and close this issue, I think.

**Comment by milanwiedemann** [comment]
I tried today setting up our venv (`just devenv` works now on M1, previously installing `pymssql` failed). also, all unit tests also work. however the spec tests (or anything that needs the mssql database) fails with various errors, here some examples:

- `
  File "/Users/milanwiedemann/Code/opensafely-core/databuilder/.venv/lib/python3.11/site-packages/_pytest/fixtures.py", line 1051, in execute
    fixturedef = request._get_active_fixturedef(argname)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
`

- `
ERROR tests/spec/sort_and_pick/test_sort_by_interleaved_with_where.py::test_sort_by_before_where[dump_sql-mssql] - ImportError: dlopen(/Users/milanwiedemann/Code/opensafely-core/databuilder/.venv/lib/python3.11/site-packages/pymssql/_mssql.cpython-311-darwin.so, 0x0002): symbol not found in flat namespace '_bcp_batch'
`

- `
docker.errors.NotFound: 404 Client Error for http+docker://localhost/v1.43/containers/c34d70009d30da360b40ac6ff45452818faf79972f3113f19431babeb47f930a/start: Not Found ("platform linux/arm64/v8 not supported")
`

**Comment by milanwiedemann** [comment]
I tried running this again with Docker Version 4.24.2 (124339) on M1 Mac, both with and without the developmental feature _Rosetta for x86/amd64 emulation_ enabled. As before, `just devenv` and `just test-unit` work but any tests that use databases (e.g., `just test-spec`) error with the following message:

`
ImportError: dlopen(/Users/milanwiedemann/Code/opensafely-core/ehrql/.venv/lib/python3.11/site-packages/pymssql/_mssql.cpython-311-darwin.so, 0x0002): symbol not found in flat namespace '_bcp_batch'
`

**Comment by suzannehamilton** [comment]
I'm running Docker version 4.41.2 (191736) on an M4 Mac. In this version of Docker Desktop, the Rosetta option is in the General menu and enabled by default:
<img width="1104" alt="Image" src="https://github.com/user-attachments/assets/63f58ae3-f251-4049-a23e-0cd5c6a85e05" />

I haven't seen any Docker issues in the tests, though I'm not sure if the tests have got far enough to actually use the containers. (Though `just test-docker` does successfully start the `ehrql-mssql` and `ehrql-trino` containers, which is a good sign.)

The test errors are similar to the psmssql ones that Milan reported.

`just test-acceptance` and `just test-docs-examples` pass without any issues.

`just test-unit`, just test-feature`, `just test-docker` and `just test-generative` fail with errors like:
```
E   ImportError: dlopen(/Users/suzanne/.pyenv/versions/3.11.13/envs/ehrql/lib/python3.11/site-packages/pymssql/_mssql.cpython-311-darwin.so, 0x0002): symbol not found in flat namespace '_bcp_batch'
```

The unit tests worked for Milan in 2023. I think the difference is [this change](2a9a93bf3), which imports some packages from pymssql to test the exception handling.
